### PR TITLE
JSON parsing: add shallow parsing up to a certain path depth

### PR DESCRIPTION
**Pull request by patrickmann**
Adds a new optional parameter to the `parse_json` pipeline function. The parameter specifies the maximum parse depth - we ignore any nodes that are more deeply nested.

See related issues for details and use case.

Resolves Graylog2/graylog2-server#9021

### CODE REVIEW

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/json/JsonUtils.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                    }
+                }
+                index++;
+            }
+            // avoid ConcurrentModificationException by delaying removal
+            int offset = 0;
+            for (int i : toRemove) {
+                ((ArrayNode) root).remove(i - offset++);
+            }
+        }
```
Comment (thll) [inline graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/json/JsonUtils.java:74]: I'm not sure, if I'm missing something but I have the feeling we could maybe simplify this. Could you please check if the following condensed version of the above would work as well?

```suggestion
        final Iterator<JsonNode> elements = root.elements();
        while (elements.hasNext()) {
            final JsonNode node = elements.next();
            if (node.isContainerNode()) {
                if (depth >= maxDepth) {
                    elements.remove();
                } else {
                    deleteBelow(node, maxDepth, depth + 1);
                }
            }
        }
```

**Review by thll — CHANGES_REQUESTED** [review]
Tested it and it worked well.

Didn't find any issues with the code but I've left some suggestions nonetheless.

File: graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/json/JsonUtils.java (hunk 1 above)
Comment (patrickmann) [reply graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/json/JsonUtils.java:74]: You are quite right - I was so focused on the custom removal methods provided by JsonObject and JsonArray that I missed the iterator's remove(). I was wondering why there was nothing at the JsonNode level. :/

**Review by thll — APPROVED** [review]
LGTM

**Pull request merged**